Evolution mortar structures

nilsdeppe opened the pull request:
## Proposed changes

- Add data structure and tags for new mortar data
- Add initialization action for new mortar data

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Add new MortarData structure for evolution

Add new structure for the mortar data in the evolution code. This will also hold
the volume values for DG-subcell once we have that integrated. Doing so will
allow reducing the number of communications between neighboring elements.

nilsdeppe committed:
Add mortar tags for evolution boundaries

nilsdeppe committed:
Add mortar initialization action

kidder commented on src/Evolution/DiscontinuousGalerkin/MortarData.hpp:
```diff
@@ -0,0 +1,89 @@
+ * is referred to as the `neighbor` data. The boundary correction schemes
+ * that store data in the `MortarData` must always store fields on the mortar
+ * (nothing else makes sense). The mesh on the interface of each neighbor is
+ * also store since the mesh will be necessary when hybridizing DG with finite
```
also stored

kidder commented on src/Evolution/DiscontinuousGalerkin/MortarData.hpp:
```diff
@@ -0,0 +1,89 @@
+template <size_t Dim>
+class MortarData {
+ public:
+  /// Add a value. This function must be called once between calls to
```
Add a value -> Insert data onto the mortar.

kidder commented on src/Evolution/DiscontinuousGalerkin/MortarData.hpp:
```diff
@@ -0,0 +1,89 @@
+  //@}
+
+  /// Return the inserted data and reset the state to empty.
+  std::pair<std::pair<Mesh<Dim - 1>, std::vector<double>>,
```
document which of the outer pair is the local/remote data

kidder commented on src/Evolution/DiscontinuousGalerkin/MortarData.hpp:
```diff
@@ -0,0 +1,89 @@
+  /// Add a value. This function must be called once between calls to
+  /// extract.
+  //@{
+  void insert_local_mortar_data(TimeStepId time_step_id, Mesh<Dim - 1> mesh,
```
mesh -> local_interface_mesh

kidder commented on src/Evolution/DiscontinuousGalerkin/MortarData.hpp:
```diff
@@ -0,0 +1,89 @@
+  //@{
+  void insert_local_mortar_data(TimeStepId time_step_id, Mesh<Dim - 1> mesh,
+                                std::vector<double> vars) noexcept;
+  void insert_remote_mortar_data(TimeStepId time_step_id, Mesh<Dim - 1> mesh,
```
mesh -> remote_interface_mesh

kidder replied:
Document the expectations of `remote_interface_mesh` and `vars` with respect to the orientation between the local logical frame and the neighbor logical frame (i.e. is the inserter or extractor supposed to worry about the orientation)

kidder commented on src/Evolution/DiscontinuousGalerkin/MortarData.hpp:
```diff
@@ -0,0 +1,89 @@
+ * (nothing else makes sense). The mesh on the interface of each neighbor is
+ * also store since the mesh will be necessary when hybridizing DG with finite
+ * difference or finite volume schemes (DG-subcell).
+ *
```
Document that data can only exist at a single TimeStepId

kidder commented on src/Evolution/DiscontinuousGalerkin/MortarData.cpp:
```diff
@@ -0,0 +1,102 @@
+         "Received local data at " << time_step_id
+                                   << ", but already have remote data at "
+                                   << time_step_id_);
+  // NOLINTNEXTLINE(performance-move-const-arg)
```
not sure if here, but somewhere assert the size of local_vars and remote_vars is the same

kidder commented on tests/Unit/Evolution/DiscontinuousGalerkin/Test_MortarTags.cpp:
```diff
@@ -0,0 +1,32 @@
+#include <cstddef>
+
+#include "DataStructures/DataBox/Tag.hpp"
+#include "DataStructures/DataVector.hpp"
```
you should only need MortarTags and TestHelpers

kidder commented on src/Evolution/DiscontinuousGalerkin/Initialization/Mortars.cpp:
```diff
@@ -0,0 +1,102 @@
+  MortarMap<evolution::dg::MortarData<Dim>> mortar_data{};
+  MortarMap<Mesh<Dim - 1>> mortar_meshes{};
+  MortarMap<std::array<Spectral::MortarSize, Dim - 1>> mortar_sizes{};
+  for (const auto& direction_and_neighbors : element.neighbors()) {
```
I like structured bindings here

nilsdeppe commented:
@kidder I've rebased on develop and pushed fixups. Thanks for the review!

kidder commented on src/Evolution/DiscontinuousGalerkin/MortarData.hpp:
```diff
@@ -20,29 +20,40 @@ namespace evolution::dg {
+ *
+ * If the element and its (remote) neighbor have unaligned logical coordinate
+ * systems then the data is stored in the local logical coordinate's
+ * orientation (xi varies fastest). This means the action sending the data is
```
optional: use latex for xi

nilsdeppe replied:
Done and squashed. Thanks!

kidder reviewed (commented):
you can squash (including one optional change)

kidder commented:
@fmahebert can you provide the second review for this?

fmahebert commented:
Yes, working on it now :)

fmahebert commented:
I need more documentation before I can understand this code. In particular, I was looking at the first commit adding MortarData.hpp/cpp, and got stuck on:
- Please clarify if there is one or two mortars between two neighboring elements. In other words, does the "local" element own its mortar to a neighbor to the right, and the neighbor own a separate mortar to the local element on its left? or is there a unique mortar performing a unique computation? 
- You use "remote" in different and surprising ways, but you've never defined the term. In the doxygen you say there is the "local data" from the element and the "neighbor data" from the neighbor, but then you call things "remote neighbor" and "remote mortar" and "remote data" and I don't know what these mean.
-  Please give clear descriptions of how `insert_local_mortar_data` and `insert_remote_mortar_data` are to be used, because the string of assertions within these functions is very hard to understand.

nilsdeppe commented:
Rebased and pushed new commit.
- renamed `remote` to `neighbor` (this is true even at external boundaries because it is a ghost neighbor)
- rewrote the entire documentation

kidder commented on src/Evolution/DiscontinuousGalerkin/MortarData.hpp:
```diff
@@ -0,0 +1,117 @@
+ * DG scheme.
+ *
+ * The class holds the local data that has been projected to the mortar as well
+ * as the neighbor data that has been projected to the neighbor. The local and
```
"projected to the mortar"

kidder commented on src/Evolution/DiscontinuousGalerkin/MortarData.hpp:
```diff
@@ -0,0 +1,117 @@
+   *   only checked if the local/neighbor data was already inserted)
+   *
+   * \note it is not required that the number of grid points between the local
+   * and neighbor data be the same since one may be used FD/FV instead of DG and
```
used -> using

kidder commented on src/Evolution/DiscontinuousGalerkin/Initialization/Mortars.hpp:
```diff
@@ -0,0 +1,134 @@
+ * If the template parameter `AddFluxBoundaryConditionMortars`
+ * is set to `false` then the mortar data for flux boundary conditions are not
+ * initialized and other boundary conditions can be applied. In this case, the
+ * `Tags::Mortars` items have no entries for external boundary directions.
```
I don't know what `Tags::Mortars` refers to as it is a PrefixTag that does not appear to be used

nilsdeppe commented:
rebased on develop and pushed fixups. Thanks for the reviews!

fmahebert commented:
I think it'll be easier for me to review this (since my first review was not at a detailed level) after a squash of the fixups. @kidder can you let @nilsdeppe know if he can squash?

kidder commented:
you can squash

nilsdeppe commented:
Okay, squashed!

fmahebert commented on src/Evolution/DiscontinuousGalerkin/MortarData.hpp:
```diff
@@ -0,0 +1,117 @@
+ * as the neighbor data that has been projected to the mortar. The local and
+ * neighbor data is later used to compute the same unique boundary correction on
+ * the mortar for both elements. That is, the final boundary correction
+ * computation is done twice: one on each element touching the mortar. However,
```
typo: one -> once (I think?)

fmahebert commented on src/Evolution/DiscontinuousGalerkin/MortarData.hpp:
```diff
@@ -0,0 +1,117 @@
+  // clang-tidy: google-runtime-references
+  void pup(PUP::er& p) noexcept;  // NOLINT
+
+  /// Retrieve the local data at `time_step_id`
```
This seems redundant with `local_mortar_data()` above. I'd suggest either removing one of them (not necessarily this one) or documenting why there are two (in other words, documenting when to use which).

nilsdeppe replied:
I honestly don't know why this exists.... I'll remove it.

fmahebert commented on src/Evolution/DiscontinuousGalerkin/MortarData.hpp:
```diff
@@ -0,0 +1,117 @@
+ public:
+  /*!
+   * \brief Insert data onto the mortar. This function must be called once
+   * between calls to extract.
```
Clarify (probably not in the brief, but below) if you mean each of the "local" and "neighbor" functions is to be called once (= 2 total calls), or only one of them is called only one time.

fmahebert commented on src/Evolution/DiscontinuousGalerkin/MortarData.hpp:
```diff
@@ -0,0 +1,117 @@
+   * between calls to extract.
+   *
+   * The insert functions require that:
+   * - the data is inserted only once
```
Same question. Is it local data once + neighbor data once? or "some data once"? Maybe clarifying here is enough.

nilsdeppe replied:
Since with the Doxygen comment is applied once to each function I'm not clear it's confusing. I did add a sentence before the bullet list `Exactly one local and neighbor insert call must be made between calls to extract`. What are your thoughts?

fmahebert replied:
To me this sentence clarified things nicely, thank you 🙂

fmahebert commented on tests/Unit/Evolution/DiscontinuousGalerkin/Test_MortarData.cpp:
```diff
@@ -0,0 +1,91 @@
+  fill_with_random_values(make_not_null(&local_data), make_not_null(&gen),
+                          make_not_null(&dist));
+  REQUIRE(local_data.size() / number_of_components ==
+          mortar_mesh.number_of_grid_points());
```
As far as I can tell, the mortar itself doesn't care if this REQUIRE is satisfied or not, right? Why is it here? If it's just a sanity check on the test setup and not a requirement on the mortar, can you perhaps add a short comment clarifying this? Something like `// Sanity-check the size of the data (but mortar doesn't check this)` or similar? Alternatively, remove.

(Make same changes in the similar REQUIRE below)

fmahebert replied:
@nilsdeppe I see you've opted to remove the first require, but you left the second one a few lines below.

fmahebert commented on src/Evolution/DiscontinuousGalerkin/Initialization/Mortars.hpp:
```diff
@@ -0,0 +1,134 @@
+                  not db::tag_is_retrievable_v<Tags::MortarMesh<Dim>,
+                                               db::DataBox<DbTagsList>> and
+                  not db::tag_is_retrievable_v<Tags::MortarSize<Dim>,
+                                               db::DataBox<DbTagsList>>) {
```
should this also check for `MortarNextTemporalId` ?

fmahebert commented on src/Evolution/DiscontinuousGalerkin/Initialization/Mortars.cpp:
```diff
@@ -0,0 +1,100 @@
+                  MortarMap<Mesh<Dim - 1>>,
+                  MortarMap<std::array<Spectral::MortarSize, Dim - 1>>,
+                  MortarMap<TimeStepId>> {
+  static constexpr size_t dim = Dim;
```
In the body of the function you often use Dim and twice use dim. I suggest to remove this and use `Dim` consistently.

fmahebert commented on src/Evolution/DiscontinuousGalerkin/Initialization/Mortars.cpp:
```diff
@@ -0,0 +1,100 @@
+  }
+
+  // We will need to update the AddFluxBoundaryConditionMortars to check the
+  // actual boundary condition being imposed.
```
This comment isn't very clear to me, because I'm confused about the concept of "update the AddFluxBoundaryConditionMortars".

I'm guessing you're talking about code changes in future PRs where the BCs will be implemented? Perhaps rephrasing to something like "In a future update, we will update the logic below to also check the actual boundary condition being imposed" would be more helpful... to me at least.

fmahebert commented on src/Evolution/DiscontinuousGalerkin/Initialization/Mortars.cpp:
```diff
@@ -0,0 +1,100 @@
+    const auto& neighbors = direction_and_neighbors.second;
+    for (const auto& neighbor : neighbors) {
+      const auto mortar_id = std::make_pair(direction, neighbor);
+      mortar_next_temporal_ids.insert({mortar_id, next_temporal_id});
```
Why is this in a separate loop over neighbors? Why not perform this insertion in the earlier loop, around line 54?

If it has to stay down here, then
- add a comment saying why
- use structured bindings `[direction, neighbors] : element.neighbors()` for consistency with the prior loops

fmahebert commented on tests/Unit/Evolution/DiscontinuousGalerkin/Initialization/Test_Mortars.cpp:
```diff
@@ -0,0 +1,293 @@
+
+template <>
+void test<3>() {
+  INFO("3D");
```
I know you can't draw ascii art here as easily, but could you at least write comments (just like in the 1d, 2d cases) to describe what the setup of the element and the mortars is?

fmahebert reviewed (commented):
Thanks for cleaning up the docs previously, this was easy to read for the most part 👍

nilsdeppe commented:
pushed a couple fixup commits. Thanks for the review!

fmahebert commented:
I'm happy with fixups, modulo the one thing I pinged you about above. Please squash everything 👍

nilsdeppe commented:
Okay, squashed. Thank you so much for the reviews!
